Source/DafnyLanguageServer/Handlers/DafnyCompletionHandler.cs

```diff
@@ -86,11 +93,81 @@ public CompletionList Process() {
         return new CompletionList();
       }
 
+      private string GetLastTwoCharactersBeforePositionIncluded(TextReader fileContent, DafnyPosition position) {
+        // To track the last two characters
+        char? prevChar = null;
+        char? currentChar = null;
+        var targetLineIndex = position.Line;
+        var targetColumnIndex = position.Column - 1;
+
+        // Read line by line
+        for (var lineNum = 0; lineNum <= targetLineIndex; lineNum++) {
+          var line = fileContent.ReadLine();
+          if (line == null) {
+            logger.LogWarning("Reached end of file before finding the specified line.");
+            return "";
+          }
+
+          // If we are on the line of the specified position, handle the partial line case
+          if (lineNum == targetLineIndex) {
+            int columnIndex = targetColumnIndex; // Convert 1-based to 0-based index
+            for (int i = 0; i <= columnIndex; i++) {
+              prevChar = currentChar;
+              if (line.Length <= i) { // In some tests (e.g. SlowlyTypeFile) the position is given only as a character index
+                targetLineIndex += 1;
+                targetColumnIndex -= line.Length + 1; // 1 for the newline. It does not need to be OS-specific, since it's just for a coverage test
+              } else {
+                currentChar = line[i];
+              }
+            }
+
+            if (lineNum < targetLineIndex) {
+              continue;
+            }
+            break;
+          } else {
+            // Otherwise, track the last two characters of this full line (including newline)
+            foreach (var c in line + '\n')  // Include '\n' for line end tracking
+            {
+              prevChar = currentChar;
+              currentChar = c;
+            }
+          }
+        }
+
+        // Handle case where fewer than 2 characters are available
+        if (prevChar == null) {
+          return currentChar?.ToString() ?? "";
+        }
+        return $"{prevChar}{currentChar}";
+      }
+
       private bool IsDotExpression() {
         var node = state.Program.FindNode<INode>(request.TextDocument.Uri.ToUri(), request.Position.ToDafnyPosition());
         return node?.RangeToken.EndToken.val == ".";
       }
 
+      private bool IsAtAttribute() {
+        var position = request.Position.ToDafnyPosition();
+        if (projectManager == null) {
+          return false;
+        }
+        var fileContent = projectManager.ReadFile(request.TextDocument.Uri.ToUri());
```

Comment:
I'll get back to this soon, but meantime, do you have a hint about where the already processed file contents is? I thought ReadFile would just give me the current string

Reply:
I've pushed a commit that addresses this comment.

Reply:
Do you know why your commit breaks some tests? For example:

Could not close file file:///tmp/z3sq5jsc.kyh/DocumentStaysUnloadedWhenClosed.dfy because it was not open.